src/stash/src/sqlite.rs

```diff
@@ -155,87 +154,103 @@ impl StashConn for Sqlite {
 
         tx.commit()?;
         Ok(StashCollection {
-            conn: Self {
-                conn: Arc::clone(&self.conn),
-            },
-            collection_id,
+            id: collection_id,
             _kv: PhantomData,
         })
     }
 
-    fn iter(
+    fn iter<K, V>(
         &self,
-        collection_id: Id,
-    ) -> Result<Vec<((Vec<u8>, Vec<u8>), Timestamp, Diff)>, StashError> {
+        collection: StashCollection<K, V>,
+    ) -> Result<Vec<((K, V), Timestamp, Diff)>, StashError>
+    where
+        K: Codec + Ord,
+        V: Codec + Ord,
+    {
         let mut conn = self.conn.lock().expect("lock poisoned");
         let tx = conn.transaction()?;
-        let since = match self.since_tx(&tx, collection_id)?.into_option() {
+        let since = match self.since_tx(&tx, collection.id)?.into_option() {
             Some(since) => since,
             None => {
                 return Err(StashError::from(
                     "cannot iterate collection with empty since frontier",
                 ));
             }
         };
-        let rows = tx
+        let mut rows = tx
             .prepare(
                 "SELECT key, value, time, diff FROM data
                  WHERE collection_id = $collection_id",
             )?
-            .query_and_then(named_params! {"$collection_id": collection_id}, |row| {
-                let key = row.get("key")?;
-                let value = row.get("value")?;
+            .query_and_then(named_params! {"$collection_id": collection.id}, |row| {
+                let key_buf: Vec<_> = row.get("key")?;
+                let value_buf: Vec<_> = row.get("value")?;
+                let key = K::decode(&key_buf)?;
```

Comment:
Hm, we can bring the previous behavior back where the backing store only deals with bytes but that will make a `persist` backed implementation awkward because `persist` expects a `(K: Codec, V: Codec)` pair and we'd be converting `K -> Vec<u8>` at some outer layer before passing them to `persist`, which would at that point only see bytes.